Add integration for Schluter

prairieapps opened the pull request:
<!--
  You are amazing! Thanks for contributing to our project!
  Please, DO NOT DELETE ANY TEXT from this template! (unless instructed).
-->


## Proposed change
<!-- 
  Describe the big picture of your changes here to communicate to the
  maintainers why we should accept this pull request. If it fixes a bug
  or resolves a feature request, be sure to link to that issue in the
  additional information section.
-->
Created a new integration for Schluter DITRA-HEAT-E-WiFi electric floor heating thermostats. Via py-schluter PyPi package, communicates with Schluter API to implement the climate platform. This is my first integration for Home Assistant.


## Type of change
<!--
  What type of change does your PR introduce to Home Assistant?
  NOTE: Please, check only 1! box! 
  If your PR requires multiple boxes to be checked, you'll most likely need to
  split it into multiple PRs. This makes things easier and faster to code review.
-->

- [ ] Dependency upgrade
- [ ] Bugfix (non-breaking change which fixes an issue)
- [x] New integration (thank you!)
- [ ] New feature (which adds functionality to an existing integration)
- [ ] Breaking change (fix/feature causing existing functionality to break)
- [ ] Code quality improvements to existing code or addition of tests

## Example entry for `configuration.yaml`:
<!--
  Supplying a configuration snippet, makes it easier for a maintainer to test
  your PR. Furthermore, for new integrations, it gives an impression of how
  the configuration would look like.
  Note: Remove this section if this PR does not have an example entry.
-->

```yaml
# Example configuration.yaml
schluter:
  username: YOUR_SCHLUTER_EMAIL
  password: YOUR_SCHLUTER_PASSWORD
```

## Additional information
<!--
  Details are important, and help maintainers processing your PR.
  Please be sure to fill out additional details, if applicable.
-->

- Link to documentation pull request: https://github.com/home-assistant/home-assistant.io/pull/11848

## Checklist
<!--
  Put an `x` in the boxes that apply. You can also fill these out after
  creating the PR. If you're unsure about any of them, don't hesitate to ask.
  We're here to help! This is simply a reminder of what we are going to look
  for before merging your code.
-->

- [x] The code change is tested and works locally.
- [x] Local tests pass. **Your PR cannot be merged unless tests pass**
- [x] There is no commented out code in this PR.
- [x] I have followed the [development checklist][dev-checklist]
- [x] The code has been formatted using Black (`black --fast homeassistant tests`)
- [ ] Tests have been added to verify that the new code works.

If user exposed functionality or configuration variables are added/changed:

- [x] Documentation added/updated for [www.home-assistant.io][docs-repository]

If the code communicates with devices, web services, or third-party tools:

- [x] The [manifest file][manifest-docs] has all fields filled out correctly.  
      Updated and included derived files by running: `python3 -m script.hassfest`.
- [x] New or updated dependencies have been added to `requirements_all.txt`.  
      Updated by running `python3 -m script.gen_requirements_all`.
- [x] Untested files have been added to `.coveragerc`.

The integration reached or maintains the following [Integration Quality Scale][quality-scale]:
<!--
  The Integration Quality Scale scores an integration on the code quality
  and user experience. Each level of the quality scale consists of a list
  of requirements. We highly recommend getting your integration scored!
-->

- [x] No score or internal
- [ ] 🥈 Silver
- [ ] 🥇 Gold
- [ ] 🏆 Platinum

<!--
  Thank you for contributing <3

  Below, some useful links you could explore:
-->
[dev-checklist]: https://developers.home-assistant.io/docs/en/development_checklist.html
[manifest-docs]: https://developers.home-assistant.io/docs/en/creating_integration_manifest.html
[quality-scale]: https://developers.home-assistant.io/docs/en/next/integration_quality_scale_index.html
[docs-repository]: https://github.com/home-assistant/home-assistant.io

prairieapps committed:
Add integration for Schluter
New integration for Schluter DITRA-HEAT thermostats. Interacts with Schluter API via py-schluter to implement climate platform.

homeassistant commented:
Hi @prairieapps, 

It seems you haven't yet signed a CLA. Please do so [here](https://home-assistant.io/developers/cla_sign_start/?pr=home-assistant&#x2F;home-assistant#31088).

Once you do that we will be able to review and accept this pull request.

Thanks!

prairieapps committed:
Fix for manifest issue on build
Removed unnecessary configurator logic

prairieapps committed:
Flake8 & ISort issues fixed

frenck commented on homeassistant/components/schluter/climate.py:
```diff
@@ -0,0 +1,171 @@
+
+    for thermostat in data.thermostats:
+        devices.append(SchluterThermostat(thermostat, temp_unit, data))
+        _LOGGER.debug("Adding thermostat: %s", thermostat.name)
```
This is already logged by the core (the creation of entities).

prairieapps replied:
I was not aware of this, thanks! Removed this unnecessary log statement.

frenck commented on homeassistant/components/schluter/climate.py:
```diff
@@ -0,0 +1,171 @@
+    @property
+    def hvac_mode(self):
+        """Return current operation ie. heat, idle."""
+        return MODE_SCHLUTER_TO_HASS[self._is_heating]
```
Do we need a lookup for this?
```suggestion
        return HVAC_MODE_HEAT if self._is_heating else HVAC_MODE_HEAT
```

prairieapps replied:
Agreed, absolutely not. I've used your suggestion (with a slight change where the "else" is HVAC_MODE_OFF, and I've removed the no longer necessary MODE_SCHLUTER_TO_HASS

frenck reviewed (commented):
Some small initial comments to get started. Could you take a look @prairieapps ? Thanks! 👍

prairieapps commented on homeassistant/components/schluter/__init__.py:
```diff
@@ -0,0 +1,178 @@
+    try:
+        api_http_session = Session()
+    except RequestException as ex:
+        _LOGGER.warning("Creating HTTP session failed with: %s", str(ex))
```
I really appreciate this code review @frenck . I believe this should be a warning as failing to create an HTTP session is not a show-stopper. I have removed the string cast though.

prairieapps committed:
Code review modifications for Schluter integration
- Removed unnecessary strings.json file
- Removed unnecessary DEFAULT_SCAN_INTERVAL from __init__.py
- Removed persistent notification for authentication error in __init__.py
- Removed string casts from log statements in __init__.py
- Removed unnecessary logging of entity creations in climate.py
- Removed unnecessary lookup for hvac_mode in climate.py

prairieapps commented on homeassistant/components/schluter/__init__.py:
```diff
@@ -0,0 +1,178 @@
+    try:
+        authentication = authenticator.authenticate()
+    except RequestException as ex:
+        _LOGGER.error("Unable to connect to Schluter service: %s", str(ex))
```
String casts removed from log statements

prairieapps commented on homeassistant/components/schluter/__init__.py:
```diff
@@ -0,0 +1,178 @@
+    except RequestException as ex:
+        _LOGGER.error("Unable to connect to Schluter service: %s", str(ex))
+
+        hass.components.persistent_notification.create(
```
Persistent notification has been removed.

prairieapps commented on homeassistant/components/schluter/__init__.py:
```diff
@@ -0,0 +1,178 @@
+from .const import DOMAIN
+
+_LOGGER = logging.getLogger(__name__)
+DEFAULT_SCAN_INTERVAL = timedelta(seconds=5)
```
Yep, unused. It is removed.

prairieapps commented on homeassistant/components/schluter/strings.json:
```diff
@@ -0,0 +1,21 @@
+{
```
Correct, no configuration flow so I have removed this file.

prairieapps committed:
Work started on Update Coordinator

prairieapps committed:
Further Coordinator work

prairieapps committed:
Update to DataUpdateCoordinator complete

prairieapps committed:
Flake8 fix

prairieapps commented:
@frenck Thanks again for the review.

I have addressed all the issues you noted. The biggest bit of work was implementing the DataUpdateCoordinator, but I much prefer it, cleans up the code quite a bit. Thanks for providing the documentation.

frenck commented on homeassistant/components/schluter/__init__.py:
```diff
@@ -0,0 +1,93 @@
+    return await async_setup_schluter(hass, config, api, authenticator)
+
+
+class SchluterPlatformData:
```
You could make this a `@dataclass` now, or... even eliminate it by storing both values in `hass.data` directly.

frenck commented on homeassistant/components/schluter/climate.py:
```diff
@@ -0,0 +1,161 @@
+        _LOGGER,
+        name="schluter",
+        update_method=async_update_data,
+        update_interval=timedelta(seconds=10),
```
```suggestion
        update_interval=SCAN_INTERVAL,
```

frenck commented on homeassistant/components/schluter/climate.py:
```diff
@@ -0,0 +1,161 @@
+    @property
+    def should_poll(self):
+        """Return if platform should poll."""
+        return True
```
Polling is not needed when using the coordinator. The entity should subscribe to changes from the data coordinator:

```py
    async def async_added_to_hass(self):
        """When entity is added to hass."""
        self.coordinator.async_add_listener(
            self.async_write_ha_state
        )

    async def async_will_remove_from_hass(self):
        """When entity will be removed from hass."""
        self.coordinator.async_remove_listener(
            self.async_write_ha_state
        )
```

See coordinator developer documentation for more information and more extensive example

prairieapps committed:
Removal of unnecessary SchluterPlatformData
Coordinator also now using SCAN_INTERVAL

prairieapps committed:
Disable polling, use coordinator listeners

prairieapps committed:
Isort on climate.py

prairieapps commented:
@frenck Thanks again for the further review.

I've pushed changes removing the unnecessary SchluterPlatformData class, the coordinator using SCAN_INTERVAL, and disabling polling and using coordinator listeners within the entity.

frenck commented on homeassistant/components/schluter/__init__.py:
```diff
@@ -0,0 +1,86 @@
+
+    if state == AuthenticationState.AUTHENTICATED:
+        hass.data[DATA_SCHLUTER_SESSION] = authentication.session_id
+        hass.data[DATA_SCHLUTER_API] = api
```
`hass.data` must be namespaced by the domain
```suggestion
        hass.data[DOMAIN] = {
            DATA_SCHLUTER_API: api,
            DATA_SCHLUTER_SESSION: authentication.session_id,
        }
```

prairieapps replied:
Thanks for the suggested change, I wasn't quite sure I was doing that properly.

frenck commented on homeassistant/components/schluter/__init__.py:
```diff
@@ -0,0 +1,86 @@
+_LOGGER = logging.getLogger(__name__)
+DATA_SCHLUTER_SESSION = "schluter_session"
+DATA_SCHLUTER_API = "schluter_api"
+PLATFORMS = ["climate"]
```
Do we get more platforms in the future? If not, we can remove this and just forward it to that single platform directly.

prairieapps replied:
My initial intention was to make it extensible to other platforms, but realistically I think climate is likely to be the only platform based on Schluter's product line. I will make this change

frenck commented on homeassistant/components/schluter/climate.py:
```diff
@@ -0,0 +1,169 @@
+    @property
+    def hvac_modes(self):
+        """List of available operation modes."""
+        return None
```
We have modes right? `havc_mode` returns some?

I guess this indicates a small issue. As in `hvac_mode` defines the mode the user has set (e.g., heating, cooling). If the device support just heating, `hvac_mode` should always return HVAC_MODE_HEAT. and `hvac_modes` should return a list with just heat as well.

If you want to indicate/show the current action the device is executing (e.g., idling, heating e.d.), the `hvac_action` should be used additionally

```py
    @property
    def hvac_action(self) -> Optional[str]:
        """Return the current running hvac operation if supported.

        Need to be one of CURRENT_HVAC_*.
        """
```

prairieapps replied:
I ended up doing most of what you suggested here. I switched the check on is_heating to use HVAC Action instead of mode. This is actually exactly what I wanted, because "Idle" is much more accurate than "Off". 

I experimented with returning HVAC_MODE_HEAT from hvac_modes, but with the nature of this particular in-floor heating thermostat, there really is no notion of switching modes. So when I return that list, a flame icon shows up on the entity UI that is clickable, and I don't really want that in this case. 

With this thermostat, essentially you set the temperature, and if it's set higher than the current temperature, it's heating, if not, it's idle. So I think the switch to HVAC Action makes sense and looks a lot better.

frenck replied:
In this case the havc mode is always heating... We should not change the backend behavior just because we don't like the frontend... that is a different problem and not a concern implementing this.

prairieapps replied:
Fair enough! I've modified the hvac_modes function to return HVAC_MODE_HEAT. This, of course, added the icon to the front end, so I added async_set_hvac_mode to avoid HA throwing a NotImplementedException when that icon is clicked.

prairieapps replied:
![Image of end result](https://imgur.com/rtiHglt)

frenck commented on homeassistant/components/schluter/climate.py:
```diff
@@ -0,0 +1,169 @@
+    @property
+    def fan_modes(self):
+        """List of available fan modes."""
+        return None
```
This entity does not have a fan in its supported features, this doesn't need to be implemented and can be removed.

prairieapps replied:
This has been removed 👍

frenck commented:
Sorry for the iterations, I'm trying to get comfortable with the reviews in the process as well.

I think the integration is coming along nicely and in indeed looks much slicker compared to the initial one. 💪

prairieapps commented:
No worries at all @frenck , it is much appreciated. This is my first integration, so I see this as an invaluable learning lesson. I agree it's tidying up nicely.

prairieapps committed:
Code review modifications
- Hass data stored under domain
- Since platform only likely to be climate, removed array, load explicitly
- Remove unnecessary fan mode implementation

prairieapps committed:
Switch to HVAC action for showing current heating status

prairieapps committed:
Isort fix

prairieapps commented:
@frenck All issues in your review have been addressed 👍

prairieapps committed:
HVAC Mode return list, set_hvac_mode function added

frenck commented:
✅ Looks fine to me @prairieapps! 👍 

Since I'm learning the review process, a second reviewer need to go over it to check if we both did right 😉

balloob commented on homeassistant/components/schluter/__init__.py:
```diff
@@ -0,0 +1,84 @@
+            {
+                vol.Required(CONF_USERNAME): cv.string,
+                vol.Required(CONF_PASSWORD): cv.string,
+                vol.Optional(CONF_TIMEOUT, default=DEFAULT_TIMEOUT): cv.positive_int,
```
Why would this be configurable ?

prairieapps replied:
I'll be honest, this being my first integration, I used a similar existing component as a reference. I've removed this option, I agree, not very useful in this case.

balloob commented on homeassistant/components/schluter/__init__.py:
```diff
@@ -0,0 +1,84 @@
+    conf = config[DOMAIN]
+    api_http_session = None
+    try:
+        api_http_session = Session()
```
Creating a session cannot fail as it's not doing any I/O.

prairieapps replied:
This try/catch has been removed 👍

balloob commented on homeassistant/components/schluter/__init__.py:
```diff
@@ -0,0 +1,84 @@
+
+CONFIG_SCHEMA = vol.Schema(
+    {
+        DOMAIN: vol.Schema(
```
```suggestion
        vol.Required(DOMAIN): vol.Schema(
```

prairieapps replied:
Change made

balloob commented on homeassistant/components/schluter/__init__.py:
```diff
@@ -0,0 +1,84 @@
+    try:
+        authentication = authenticator.authenticate()
+    except RequestException as ex:
+        _LOGGER.error("Unable to connect to Schluter service: %s", ex)
```
I think that you need to add a `return` here ? Because when it raises an exception, `authentication` will never be set and so line 46 will fail.

prairieapps replied:
Yes, definitely. I've added the return.

balloob commented on homeassistant/components/schluter/__init__.py:
```diff
@@ -0,0 +1,84 @@
+        _LOGGER.error("Invalid email provided")
+        return False
+
+    return False
```
Let's also log here that we are failing because of an unknown state.
```suggestion
    _LOGGER.error("Unknown set up error: %s", state)
    return False
```

prairieapps replied:
Log added for unknown state

balloob replied:
FYI, you could have clicked on "commit suggestion" in my comment and it would have been added to your code 👍

balloob commented on homeassistant/components/schluter/__init__.py:
```diff
@@ -0,0 +1,84 @@
+)
+
+
+async def async_setup_schluter(hass, config, api, authenticator):
```
You have marked this as an async function but you are calling sync functions. Please not mark this as async.

prairieapps replied:
Async has been removed from this function (and function name modified)

balloob commented on homeassistant/components/schluter/__init__.py:
```diff
@@ -0,0 +1,84 @@
+    return False
+
+
+async def async_setup(hass, config):
```
Since youre calling things that do I/O, mark this as sync version of setup.
```suggestion
def setup(hass, config):
```

prairieapps replied:
Async has been removed from this function (and function name modified)

balloob commented on homeassistant/components/schluter/__init__.py:
```diff
@@ -0,0 +1,84 @@
+        session_id_cache_file=hass.config.path(SCHLUTER_CONFIG_FILE),
+    )
+
+    return await async_setup_schluter(hass, config, api, authenticator)
```
Is there any reason you have extracted this into a new function  ?

prairieapps replied:
This file was quite a bit larger before the first code review. I've combined the functions now :)

balloob commented on homeassistant/components/schluter/climate.py:
```diff
@@ -0,0 +1,165 @@
+
+_LOGGER = logging.getLogger(__name__)
+SCAN_INTERVAL = timedelta(seconds=5)
+PLATFORM_SCHEMA = PLATFORM_SCHEMA.extend(
```
The default platform schema that climate uses already contains a scan interval. No need to define it.

prairieapps replied:
SCAN_INTERVAL is now pulled from the default platform schema 👍

balloob replied:
if you do that, it means that you don't need to define it below. It looks like you can remove all `PLATFORM_SCHEMA` related constant.

MartinHjelmare replied:
Please remove the `PLATFORM_SCHEMA` completely.

balloob commented on homeassistant/components/schluter/climate.py:
```diff
@@ -0,0 +1,165 @@
+
+    async def async_update_data():
+        try:
+            return api.get_thermostats(session_id) or []
```
You cannot call sync functions from the event loop. The update coordinator requires an async function so in this case you need to delegate the sync function to the executor.
```suggestion
            return (await hass.async_add_executor_job(api.get_thermostats, session_id)) or []
```

prairieapps replied:
Great to know, thanks. I've delegated the sync function to the executor.

balloob commented on homeassistant/components/schluter/climate.py:
```diff
@@ -0,0 +1,165 @@
+
+    async_add_entities(
+        SchluterThermostat(coordinator, idx, temp_unit, api, session_id)
+        for idx, ent in enumerate(coordinator.data)
```
This seems dangerous because if the user would add/remove a thermostat, the index is no longer correct.

I would suggest that inside your async_update_data you turn your data into a dictionary keyed by serial number.

```python
        try:
            thermostats = await hass.async_add_executor_job(api.get_thermostats, session_id)
        except RequestException as err:
            raise UpdateFailed(f"Error communicating with Schluter API: {err}")

		if thermostats is None:
            return {}

        return {thermo.serial_number: thermo for thermo in thermostats}
```

prairieapps replied:
Good catch, I agree. I've turned the data into a dictionary and made the necessary modifications in the platform to use serial number as the unique identifier & index.

balloob commented on homeassistant/components/schluter/climate.py:
```diff
@@ -0,0 +1,165 @@
+        return self.coordinator.data[self.idx].serial_number
+
+    @property
+    def device_info(self):
```
Device info is only used if you add a config flow. Easy to do in your case as it's just username/password with `python3 -m script.scaffold config_flow`. But I suggest you leave that for your next contribution :) For now let's just remove this property.

prairieapps replied:
Agreed :) I'd love to tackle config flow as a next contribution. Property removed for now.

balloob commented on homeassistant/components/schluter/climate.py:
```diff
@@ -0,0 +1,165 @@
+                self._api.set_temperature(self._session_id, serial_number, target_temp)
+        except RequestException as ex:
+            _LOGGER.error("An error occurred while setting temperature: %s", ex)
+            self.schedule_update_ha_state(True)
```
Since your entity doesn't poll, you can directly call the write state function. However, what will change that causes data in the coordinator to change ? 

```suggestion
            self.async_write_ha_state()
```

prairieapps replied:
Based on a coordinator tutorial I was using, I thought this was necessary. After some testing I think you are right, I don't think it is necessary. I've removed it

balloob commented on homeassistant/components/schluter/climate.py:
```diff
@@ -0,0 +1,165 @@
+    )
+
+
+class SchluterThermostat(ClimateDevice):
```
You can add an `available` property based on if the last request to the coordinator was successful.

```python
    @property
    def available(self):
        """Return if thermostat is available."""
        return self.coordinator.last_update_success
```

prairieapps replied:
Property added 👍

prairieapps committed:
__init__.py modifications from code review

prairieapps committed:
Climate.py code review modifications
- Device info property removed
- Write state function on set temp failure
- Add "available" property
- Delegate sync function in coordinator update to the executor

prairieapps committed:
Serial number as unique identifier
- Now using serial number & thermostat dictionary to protect against scenarios where a device is removed and enumerable index is no longer accurate
- Removed async_write_ha_state() from set temp exception, not needed
- Removed unnecessary SCAN_INTERVAL definition

prairieapps committed:
Linting

prairieapps commented:
@balloob I really appreciate the code review. Thanks! I believe I have addressed all the issues you have noted 👍

balloob commented on homeassistant/components/schluter/__init__.py:
```diff
@@ -0,0 +1,72 @@
+        api,
+        conf.get(CONF_USERNAME),
+        conf.get(CONF_PASSWORD),
+        session_id_cache_file=hass.config.path(SCHLUTER_CONFIG_FILE),
```
As a note for the future, it would be great if we don't have to pass a cache file to the authenticator and instead are able to store it in the config entry data once a config flow is added.

prairieapps replied:
Noted 👍

balloob commented on homeassistant/components/schluter/__init__.py:
```diff
@@ -0,0 +1,72 @@
+
+    conf = config[DOMAIN]
+    api_http_session = Session()
+    api = Api(timeout=conf.get(CONF_TIMEOUT), http_session=api_http_session)
```
CONF_TIMEOUT is no longer there.

prairieapps replied:
Fixed, now uses API_TIMEOUT variable

balloob commented on homeassistant/components/schluter/climate.py:
```diff
@@ -0,0 +1,167 @@
+
+
+async def async_setup_platform(hass, config, async_add_entities, discovery_info=None):
+    """Set up the Schluter thermostats."""
```
Make sure to return if discovery info is not None (someone accidentally added a platform config for it)

```suggestion
async def async_setup_platform(hass, config, async_add_entities, discovery_info=None):
    """Set up the Schluter thermostats."""
    if discovery_info is not None:
        return
```

prairieapps replied:
Merged this, needed a slight change to the code as I was passing in {} for the discovery parameter in load_platform. Thanks

balloob replied:
Yeah that is fine because `{}` is not None.

prairieapps replied:
Hi @balloob , my apologies if I am failing to understand. With your code suggestion, {} would return because it is not none. Is there a chance that the conditional should be 
`if discovery_info is None:` instead of `if discovery_info is not None:`

I was looking at a few other existing components and they seem to use
`if discovery_info is None:`

For instance, line 17 in this file
https://github.com/home-assistant/core/blob/206f8cef5c47c154354c0196b256b5feedf9478f/homeassistant/components/mychevy/binary_sensor.py

balloob commented on homeassistant/components/schluter/climate.py:
```diff
@@ -0,0 +1,167 @@
+
+    async_add_entities(
+        SchluterThermostat(coordinator, serial_number, temp_unit, api, session_id)
+        for idx, (serial_number, thermostat) in enumerate(coordinator.data.items())
```
No longer necessary to enumerate and keep track of `idx`.
```suggestion
        for serial_number, thermostat in coordinator.data.items()
```

prairieapps committed:
Update homeassistant/components/schluter/climate.py

Co-Authored-By: Paulus Schoutsen <paulus@home-assistant.io>

prairieapps committed:
Update homeassistant/components/schluter/climate.py

Co-Authored-By: Paulus Schoutsen <paulus@home-assistant.io>

prairieapps committed:
Changed timeout from config, load platform discover to None

balloob commented on homeassistant/components/schluter/__init__.py:
```diff
@@ -0,0 +1,73 @@
+            DATA_SCHLUTER_API: api,
+            DATA_SCHLUTER_SESSION: authentication.session_id,
+        }
+        discovery.load_platform(hass, "climate", DOMAIN, None, config)
```
Now that you pass `None` as discovery info, it's going to be filtered out in `async_setup_platform`. That's not good :)

prairieapps committed:
Proposed change for discover info issue

balloob reviewed (approved):
🎉

MartinHjelmare commented on homeassistant/components/schluter/climate.py:
```diff
@@ -0,0 +1,169 @@
+
+    def set_temperature(self, **kwargs):
+        """Set new target temperature."""
+        target_temp = None
```
We can remove this line.

MartinHjelmare commented on homeassistant/components/schluter/climate.py:
```diff
@@ -0,0 +1,169 @@
+        _LOGGER.debug("Setting thermostat temperature: %s", target_temp)
+
+        try:
+            if target_temp is not None:
```
Please move this check outside the try... except, invert it and return if true.

We want try... except block scope to be minimal.

MartinHjelmare commented on homeassistant/components/schluter/climate.py:
```diff
@@ -0,0 +1,169 @@
+
+    async def async_will_remove_from_hass(self):
+        """When entity will be removed from hass."""
+        self._coordinator.async_remove_listener(self.async_write_ha_state)
```
If we also implement the `async_update` entity method and await `coordinator.async_request_refresh` we allow the user to schedule manual updates via the common entity update service.

MartinHjelmare reviewed (commented):
Please address the comments in a new PR.
